share/search/index_strategy/trovesearch_denorm.py

```diff
@@ -0,0 +1,985 @@
+                'path_match': '*.text_by_propertypath.*',
+                'mapping': ts.TEXT_MAPPING,
+            }},
+            {'dynamic_text_by_depth': {
+                'path_match': '*.text_by_depth.*',
+                'mapping': ts.TEXT_MAPPING,
+            }},
+            {'dynamic_date_by_propertypath': {
+                'path_match': '*.date_by_propertypath.*',
+                'mapping': {
+                    'type': 'date',
+                    'format': 'strict_date_optional_time',
+                },
+            }},
+            {'dynamic_int_by_propertypath': {
+                'path_match': '*.int_by_propertypath.*',
+                'mapping': {'type': 'long'},
+            }},
+        ]
+
+    def _card_mappings(self):
+        return {
+            # simple keyword properties
+            'card_iri': ts.KEYWORD_MAPPING,
+            'card_pk': ts.KEYWORD_MAPPING,
+            'suid': {
+                'type': 'object',
+                'properties': {
+                    'source_config_label': ts.KEYWORD_MAPPING,
+                    'source_record_identifier': ts.KEYWORD_MAPPING,
+                },
+            },
+            **self._paths_and_values_mappings(),
+        }
+
+    def _iri_value_mappings(self):
+        return {
+            'value_iri': ts.KEYWORD_MAPPING,
+            'value_name': ts.KEYWORD_MAPPING,
+            'value_title': ts.KEYWORD_MAPPING,
+            'value_label': ts.KEYWORD_MAPPING,
+            'at_card_propertypaths': ts.KEYWORD_MAPPING,
+            **self._paths_and_values_mappings(),
+        }
+
+    def _paths_and_values_mappings(self):
+        return {
+            'focus_iri': ts.IRI_KEYWORD_MAPPING,
+            'propertypaths_present': ts.KEYWORD_MAPPING,
+            # flattened properties (dynamic sub-properties with keyword values)
+            'iri_by_propertypath': ts.FLATTENED_MAPPING,
+            'iri_by_depth': ts.FLATTENED_MAPPING,
+            # dynamic properties (see `_dynamic_templates`)
+            'text_by_propertypath': {'type': 'object', 'dynamic': True},
+            'text_by_depth': {'type': 'object', 'dynamic': True},
+            'date_by_propertypath': {'type': 'object', 'dynamic': True},
+            'int_by_propertypath': {'type': 'object', 'dynamic': True},
+        }
+
+    # override method from Elastic8IndexStrategy
+    def before_chunk(self, messages_chunk: messages.MessagesChunk, indexnames: Iterable[str]):
+        # delete all per-value docs (to account for missing values)
+        self.es8_client.delete_by_query(
+            index=list(indexnames),
+            query={'bool': {'must': [
+                {'terms': {'card.card_pk': messages_chunk.target_ids_chunk}},
+                {'exists': {'field': 'iri_value.value_iri'}},
+            ]}},
+        )
+        # (possible optimization: instead, hold onto doc_ids and (in `after_chunk`?)
+        #                         delete_by_query excluding those)
```

Comment:
For my own understanding, can you explain what/why this is doing? When messages come in chunked (e.g. for an index backfill), this will delete existing PKs from that incoming chunk if there's a `value_iri`?

Reply:
the `denorm` index has a document for the card itself and additional documents for each iri value in the card contents (with the `card` object duplicated across all of them, so a value-search can filter by anything on the card)
```
{card: ...}
{card: ..., iri_value: ...}
{card: ..., iri_value: ...}
{card: ..., iri_value: ...}
...
```
after indexed like that, if a subsequent update is missing any previous `iri_value` (say, a subject or contributor is removed), the corresponding value docs need to be deleted -- the easy/naive path was this `delete_by_query` to drop all value docs for the chunk of cards that are about to be indexed

(note it's not just for backfills -- the urgent queue is chunked as well, tho you'll more often get chunks of size one)